File: 276-paint-fence/attempt_01_bottom_up.py
# bottom-up
class Solution:
    def numWays(self, n: int, k: int) -> int:
        # handle base case
        if n == 1:
            return k
        # initialize memo array and populate base cases
        two_posts_back = k
        one_posts_back = k * k
        # iterate through number of fence posts
        for i in range(2, n, 1):
            # one way to choose same color, k ways to color last two same
            same_color = 1 * (k - 1) * two_posts_back
            # k ways to choose different color, last two posts don't matter
            different_color = (k - 1) * one_posts_back
            # shift remembered posts forward by a post
            two_posts_back = one_posts_back
            one_posts_back = same_color + different_color
        return one_posts_back

File: 769-min-cost-climbing-stairs/attempt_01_bottom_up.js
/**
 * @param {number[]} cost
 * @return {number}
 */
// bottom up
var minCostClimbingStairs = function(cost) {
    // handle base cases
    if (cost.length < 3) return Math.min(cost[0], cost[1]);
    // loop through, building memo object
    const dp = [0, 0];
    for (let i = 2; i <= cost.length; i++) {
        dp[i] = Math.min(cost[i - 1] + dp[i - 1], cost[i - 2] + dp[i - 2]);
    }
    return dp[cost.length];
};